src/ntfsck.c

```diff
@@ -2002,118 +2002,83 @@ static int ntfsck_check_file(ntfs_inode *ni)
-			check_failed("inode(%llu)'s $STD_INFO flag(%d) $DATA flag(%d)"
-					" does not set SPARSE or COMPRESSED",
-					(unsigned long long)ni->mft_no,
-					si_flags, na->data_flags);
-
-			/* if comression_block_size is not zero, attribute is comressed */
 			if (ntfsck_ask_repair(vol)) {
-				if (na->compression_block_size) {
-					si_flags &= ~FILE_ATTR_SPARSE_FILE;
-					ni->flags = si_flags |= FILE_ATTR_COMPRESSED;
-					na->data_flags &= ~ATTR_IS_SPARSE;
-					na->data_flags |= ATTR_IS_COMPRESSED;
-				} else {
-					si_flags &= ~FILE_ATTR_COMPRESSED;
-					ni->flags = si_flags |= FILE_ATTR_SPARSE_FILE;
-					na->data_flags &= ~ATTR_IS_COMPRESSED;
-					na->data_flags |= ATTR_IS_SPARSE;
-				}
-				ntfs_inode_mark_dirty(ni);
-				NInoFileNameSetDirty(ni);
-				fsck_err_fixed();
-			}
-		}
+				si_flags &= ~FILE_ATTR_COMPRESSED;
+				si_flags |= FILE_ATTR_SPARSE_FILE;
+				ni->flags = si_flags;
 
-		/* check size */
-		if ((ni->allocated_size != na->compressed_size) ||
-				(na->compressed_size != rls.real_asize)) {
-			/* need to set allocated_size & highest_vcn set */
-			check_failed("Corrupted inode(%llu) compressed size field.\n "
-					" inode allocated size(%llu),"
-					" $DATA compressed(%llu)"
-					" cluster run real allocation(%llu).",
-					(unsigned long long)ni->mft_no,
-					(unsigned long long)ni->allocated_size,
-					(unsigned long long)na->compressed_size,
-					(unsigned long long)rls.real_asize);
-			if (ntfsck_ask_repair(vol)) {
-				na->compressed_size = rls.real_asize;
-				ni->allocated_size = na->compressed_size;
 				ntfs_inode_mark_dirty(ni);
 				NInoFileNameSetDirty(ni);
 				fsck_err_fixed();
 			}
 		}
-	} else {
-		if ((si_flags & (FILE_ATTR_SPARSE_FILE | FILE_ATTR_COMPRESSED)) ||
-			(na->data_flags & (ATTR_IS_SPARSE | ATTR_IS_COMPRESSED))) {
-			check_failed("inode(%llu)'s $STD_INFO flag(%d) $DATA flag(%d)"
-					" set SPARSE or COMPRESSED",
-					(unsigned long long)ni->mft_no,
-					si_flags, na->data_flags);
-
-			/* if comression_block_size is not zero, attribute is comressed */
+	} else if (cs_flags & ATTR_IS_COMPRESSED) {
+		if (!(si_cs_flags & FILE_ATTR_COMPRESSED)) {
+			check_failed("Compressed flags of $STD_INFO and DATA in inode(%llu) "
+					"are different. Set compressed to inode.",
+					(unsigned long long)ni->mft_no);
+
 			if (ntfsck_ask_repair(vol)) {
-				si_flags &= ~(FILE_ATTR_COMPRESSED | FILE_ATTR_SPARSE_FILE);
+				si_flags &= ~FILE_ATTR_SPARSE_FILE;
+				si_flags |= FILE_ATTR_COMPRESSED;
 				ni->flags = si_flags;
-				na->data_flags &= ~(ATTR_IS_COMPRESSED | ATTR_IS_SPARSE);
 
 				ntfs_inode_mark_dirty(ni);
 				NInoFileNameSetDirty(ni);
 				fsck_err_fixed();
 			}
 		}
+	} else if (!cs_flags) {
+		if (si_cs_flags) {
+			check_failed("Sparse/Compressed flags of $STD_INFO and DATA in inode(%llu) "
+					"are different. Clear flag of inode.",
+					(unsigned long long)ni->mft_no);
 
-
-		/* check size */
-		if ((ni->allocated_size != na->allocated_size) ||
```

Comment:
you remove sizes checks, where are they checked ?

Reply:
I thought it was not necessary because related size fields modified in ntfs_attr_update_mapping_pairs(), but that is only in case of corrupted attribute. 
so I'll restore it.